test/services/database/challenge_repository_test.dart

```diff
@@ -0,0 +1,206 @@
+
+void main() {
+  late FakeFirebaseFirestore firestore;
+  late PostRepositoryService postRepository;
+  late ChallengeRepositoryService challengeRepository;
+
+  setUp(() async {
+    firestore = FakeFirebaseFirestore();
+    postRepository = PostRepositoryService(firestore: firestore);
+    challengeRepository = ChallengeRepositoryService(
+      firestore: firestore,
+      postRepositoryService: postRepository,
+    );
+  });
+
+  const userPos = userPosition1;
+  final inChallengeRange = GeoPointGenerator.createOnEdgeInsidePosition(
+    userPos,
+    ChallengeRepositoryService.maxChallengeRadius,
+  );
+
+  final uid = testingUserFirestoreId;
+
+  group("ChallengeRepositoryService", () {
+    test("Get new challenges", () async {
+      final fakePosts = await addPosts(
+        postRepository,
+        inChallengeRange,
+        ChallengeRepositoryService.maxActiveChallenges,
+      );
+
+      final challenges = await challengeRepository.getChallenges(uid, userPos);
+
+      expect(challenges.length, ChallengeRepositoryService.maxActiveChallenges);
+      for (final challenge in challenges) {
+        expect(challenge.data.isCompleted, false);
+        expect(challenge.data.expiresOn.compareTo(Timestamp.now()), isPositive);
+
+        final actualPost = await postRepository.getPost(challenge.postId);
+        expect(fakePosts.contains(actualPost.data), true);
+      }
+    });
+
+    test("Complete a challenge", () async {
+      await addPosts(postRepository, inChallengeRange, 1);
+      final challenges = await challengeRepository.getChallenges(uid, userPos);
+
+      final challenge = challenges.first;
+      expect(challenge.data.isCompleted, false);
+      await challengeRepository.completeChallenge(
+        uid,
+        challenge.postId,
+      );
+
+      final updatedChallenges =
+          await challengeRepository.getChallenges(uid, userPos);
+      expect(updatedChallenges.length, 1);
+      expect(updatedChallenges.first.data.isCompleted, true);
+    });
+
+    test("Multiple gets with only one challenge available", () async {
+      await addPosts(postRepository, inChallengeRange, 1);
+
+      for (int i = 0; i < 10; i++) {
+        final challenges =
+            await challengeRepository.getChallenges(uid, userPos);
+
+        expect(challenges.length, 1);
+      }
+    });
+
+    test("Challenge posts are unique", () async {
+      await addPosts(
+        postRepository,
+        inChallengeRange,
+        ChallengeRepositoryService.maxActiveChallenges,
+      );
+
+      final challenges = await challengeRepository.getChallenges(
+        uid,
+        userPos,
+      );
+      final postIds = challenges.map((e) => e.postId).toSet();
+      expect(postIds.length, challenges.length);
+    });
+
+    test("Deleted post disappears from challenges", () async {
+      final fakePosts = PostDataGenerator.generatePostData(1);
+      final postId =
+          await postRepository.addPost(fakePosts.first, inChallengeRange);
```

Comment:
why not using `addPosts` with `n=1`? Moreover, I think using `n=2` would make more sense: you verify that `deletePost` deletes one post, but not all of them (and maybe also check it deletes the correct one).

Reply:
You're right, I was being a bit lazy because I first wrote addPosts to only return the post data and not the ids. I added another method to return a list of PostFirestore after adding n. It should be easier to write more tests now.
73a45e83d9c53307dbb6ba9ce6f79faf301ebf90

I changed the test to use two posts as well. 3dcc90c25d88e9e705ee15990b44d291c1490966